src/server/lock_manager/deadlock.rs

```diff
@@ -110,16 +155,33 @@ impl Locks {
     }
 }
 
+type KeyReverseMap = HashMap<Vec<u8>, HashMap<TimeStamp, HashSet<TimeStamp>>>;
+
 /// Used to detect the deadlock of wait-for-lock in the cluster.
 pub struct DetectTable {
     /// Keeps the DAG of wait-for-lock. Every edge from `txn_ts` to `lock_ts`
     /// has a survival time -- `ttl`. When checking the deadlock, if the ttl
     /// has elpased, the corresponding edge will be removed.
     /// `last_detect_time` is the start time of the edge. `Detect` requests will
     /// refresh it.
-    // txn_ts => (lock_ts => Locks)
+    // waiting_txn_ts => (lock_ts => Locks)
     wait_for_map: HashMap<TimeStamp, HashMap<TimeStamp, Locks>>,
 
+    /// For finding lock-waiting and lock-holding transactions by key.
+    ///
+    /// Theoretically, keys occurring in `wait_for_map` should be unique.
+    /// However, considering that in some rare cases, it's possible that the
+    /// leader leader of some region is transferring between nodes; with
+    /// uncertain network latency, different lock-waiting relationship on a
+    /// same key can be received at the same time, while the detector leader has
+    /// no way to determine which one is the latest. Considering this, if
+    /// `key_reverse_map` is fully synced with `wait_for_map`, then a key is
+    /// possible to map to multiple existing locks, each of which has its
+    /// own set of waiting transactions.
+    // key => (lock_ts => {waiting_txn_ts})
+    // TODO: Use the pre-calculated lock hash instead to reduce the overhead of hash calculation.
```

Comment:
I didn't plan to do this in the same PR as it would be hard and complicated to make this change on the current code which already passes the hash everywhere..